### DISCUSSION TITLE
Authenticated request with custom check on X-Hasura-User-ID not working

### DISCUSSION THREAD

**Discussion by valentin-harrang** [hasura/graphql-engine#9244] [Question] [answered]
Hi,

I trying to do an authenticated request with `custom check` on `X-Hasura-User-ID` from `Next.js` App:

![image](https://user-images.githubusercontent.com/19478137/203293014-9d596f9d-14ea-4f8b-8f19-74de672e27da.png)

I get a `200` status from `Network` tab of `DevTools` after executing my `mutation`:

Mutation : 

```
const LOVE_MUTATION = gql`
  mutation insert_loved_language($objects: [loved_language_insert_input!]!) {
    insert_loved_language(objects: $objects) {
      affected_rows
    }
  }
`;

const [love, { loading }] = useMutation(LOVE_MUTATION, {
    variables: {
      objects: {
        name: "Typescript",
      },
    },
  });

```

But I get this error:

```
{"errors":[{"extensions":{"code":"validation-failed","path":"$.selectionSet.insert_loved_language.args.objects[0].name"},"message":"field 'name' not found in type: 'loved_language_insert_input'"}]}
```

Yet the `name` field does exist and the `X-Hasura-User-Id` is passed in the `payload` of my `jwt`. 

JWT:

```
jwt {
  sub: 'K0RmvMupciVEiJVwhikj',
  name: 'Valentin',
  email: 'test@gmail.com',
  iat: 1669105034.515,
  exp: 1669191434,
  'https://hasura.io/jwt/claims': {
    'x-hasura-allowed-roles': [ 'user' ],
    'x-hasura-default-role': 'user',
    'x-hasura-role': 'user',
    'x-hasura-user-id': 'K0RmvMupciVEiJVwhikj'
  }
}
```

The mutation works using the Hasura console with the `x-hasura-admin-secret` but not with `x-hasura-user-id`:

![image](https://user-images.githubusercontent.com/19478137/203293255-63e724b0-0a6f-4714-9b7c-d9a1cff04527.png)

I don't know why I have this error, do you have any idea how to fix it? 🙏

**Comment by meetzaveri** [comment, marked as answer]
Hey @valentin-harrang 

Thanks for asking. 

Since `loved_language_insert_input` schema is object wrapped around array structure, you'll need to pass like

```js
objects: [ 
      // inside array structure
      {
          name: "Typescript",     // pass your actual object that you want to insert  (which matches schema)
      }
]
```

Please let me know if that doesn't work for you!

**UPDATE (post-solved)**:
When you add column preset for column, then while performing insert mutation if you are passing value to that field via query variables, it will only accept the value which had been set in column preset. Hence, introspection query in graphiQL will show you validation failed error because that column has already been added in column preset in `INSERT` permission for that specific role (in this case `user`)

If you intend to pass value to that field from insert mutation, then please remove that column from Column presets.

**Reply by valentin-harrang** [reply]
Thank you @meetzaveri for the answer, but I have this error:

> Type "loved_language_insert_input" does not have a field "name"

![image](https://user-images.githubusercontent.com/19478137/203300474-f99edbc9-9daf-489d-a6c7-0cdca3b0e0e9.png)

And the documentation of `loved_language_insert_input`:

![image](https://user-images.githubusercontent.com/19478137/203301356-09674aea-1c69-43e6-83b7-56a8f90c5d1d.png)

I think the permissions are correct though:

![image](https://user-images.githubusercontent.com/19478137/203300989-d9c6afe3-880b-4876-9e22-b4f8b156bf1f.png)

**Reply by meetzaveri** [reply]
@valentin-harrang Could you try adding `SELECT` permissions for role `user` and in that for column select permissions, please check only `name` ✅  . Please try this and let me know if that error still persists.

**Reply by valentin-harrang** [reply]
@meetzaveri like this? 
![image](https://user-images.githubusercontent.com/19478137/203301712-a4357394-4966-454d-b0e9-effad367a313.png)

**Reply by meetzaveri** [reply]
Yes, right

**Reply by valentin-harrang** [reply]
The `select` query works: 

![image](https://user-images.githubusercontent.com/19478137/203302051-dcf0dee7-d903-4939-97e4-233c33c56038.png)

**Reply by meetzaveri** [reply]
@valentin-harrang 
This error that you are getting in insert operation is due to the reason being you have your column preset of `name` already applied with static value. If you remove that, then mutation will take the value from the query variables.

Can you try removing column preset for `name` column ? then it will accept input from query variables in insert operation.

**Reply by valentin-harrang** [reply]
@meetzaveri I can't because when I remove the column, the `Save Permissions` button is disabled

![image](https://user-images.githubusercontent.com/19478137/203304439-cbaf6cc0-0280-417f-832b-187e4e0bd3e5.png)

**Reply by meetzaveri** [reply]
Then please delete permissions and re-create it again with the solution mentioned above ( do not include `name` in column presets)

**Reply by valentin-harrang** [reply]
It works ! Thank you @meetzaveri :)
